aws/resource_aws_cloudwatch_dashboard_test.go

```diff
@@ -0,0 +1,198 @@
+					resource.TestCheckResourceAttr("aws_cloudwatch_dashboard.foobar", "dashboard_name", testAccAWSCloudWatchDashboardName(rInt)),
+				),
+			},
+		},
+	})
+}
+
+func TestAccAWSCloudWatchDashboard_update(t *testing.T) {
+	var dashboard cloudwatch.GetDashboardOutput
+	rInt := acctest.RandInt()
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckAWSCloudWatchDashboardDestroy,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSCloudWatchDashboardConfig(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudWatchDashboardExists("aws_cloudwatch_dashboard.foobar", &dashboard),
+					testAccCloudWatchCheckDashboardBodyIsExpected("aws_cloudwatch_dashboard.foobar", basicWidget),
+					resource.TestCheckResourceAttr("aws_cloudwatch_dashboard.foobar", "dashboard_name", testAccAWSCloudWatchDashboardName(rInt)),
+				),
+			},
+			{
+				Config: testAccAWSCloudWatchDashboardConfig_updateBody(rInt),
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckCloudWatchDashboardExists("aws_cloudwatch_dashboard.foobar", &dashboard),
+					testAccCloudWatchCheckDashboardBodyIsExpected("aws_cloudwatch_dashboard.foobar", updatedWidget),
+					resource.TestCheckResourceAttr("aws_cloudwatch_dashboard.foobar", "dashboard_name", testAccAWSCloudWatchDashboardName(rInt)),
+				),
+			},
+		},
+	})
+}
+
+func testAccCheckCloudWatchDashboardExists(n string, dashboard *cloudwatch.GetDashboardOutput) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		rs, ok := s.RootModule().Resources[n]
+		if !ok {
+			return fmt.Errorf("Not found: %s", n)
+		}
+
+		conn := testAccProvider.Meta().(*AWSClient).cloudwatchconn
+		params := cloudwatch.GetDashboardInput{
+			DashboardName: aws.String(rs.Primary.ID),
+		}
+
+		resp, err := conn.GetDashboard(&params)
+		if err != nil {
+			return err
+		}
+
+		*dashboard = *resp
+
+		return nil
+	}
+}
+
+func testAccCheckAWSCloudWatchDashboardDestroy(s *terraform.State) error {
+	conn := testAccProvider.Meta().(*AWSClient).cloudwatchconn
+
+	for _, rs := range s.RootModule().Resources {
+		if rs.Type != "aws_cloudwatch_dashboard" {
+			continue
+		}
+
+		params := cloudwatch.GetDashboardInput{
+			DashboardName: aws.String(rs.Primary.ID),
+		}
+
+		if _, err := conn.GetDashboard(&params); err == nil {
+			return fmt.Errorf("Dashboard still exists: %s", rs.Primary.ID)
+		} else if !isResourceNotFoundErr(err) {
+			return err
+		}
+	}
+
+	return nil
+}
+
+const (
+	basicWidget = `{
+  "widgets": [{
+    "type": "text",
+    "x": 0,
+    "y": 0,
+    "width": 6,
+    "height": 6,
+    "properties": {
+      "markdown": "Hi there from Terraform: CloudWatch"
+    }
+  }]
+}`
+
+	updatedWidget = `{
+  "widgets": [{
+    "type": "text",
+    "x": 0,
+    "y": 0,
+    "width": 6,
+    "height": 6,
+    "properties": {
+      "markdown": "Hi there from Terraform: CloudWatch - updated"
+    }
+  }]
+}`
+)
```

Comment:
I left it out so that it could be defined once and used multiple times instead of duplicating it multiple times. This also makes updating the test much easier.